packages/helldoc-core/src/app/components/Search.tsx

```diff
@@ -1,8 +1,155 @@
 import * as React from "react";
 
-class Search extends React.PureComponent {
+import { Link } from "react-router-dom";
+import siteData from "@internal/site-data";
+
+interface Result {
+  link: string;
+  mainTitle: string;
+  subTitle?: string;
+}
+
+class Search extends React.Component<
+  {},
+  {
+    fullWidth: boolean;
+    keyword: string;
+    results: Array<Result>;
+    onFocus: boolean;
+  }
+> {
+  private inputRef: React.RefObject<HTMLInputElement>;
+  constructor(props: any) {
+    super(props);
+    this.state = {
+      fullWidth: false,
+      onFocus: false,
+      keyword: "",
+      results: [
+        {
+          link: "/hell",
+          mainTitle: "Get Start",
+          subTitle: "快速开始"
+        },
+        {
+          link: "/hell/1",
+          mainTitle: "Get Start",
+          subTitle: "快速开始"
+        }
+      ]
+    };
+    this.inputRef = React.createRef();
+  }
+
+  isFullWidth() {
+    if (window.innerWidth <= 545 && !this.state.fullWidth) {
+      this.setState({ fullWidth: true });
+      const node = this.inputRef.current;
+      if (node) {
+        node.focus();
+      }
+    }
+  }
+
+  miniWidth() {
+    if (window.innerWidth <= 545) {
+      this.setState({ fullWidth: false });
+    }
+  }
+
+  searching() {
+    const searchInHeaders = () => {
+      return siteData.pages.map(({ headers, path }) => {
+        let toReturn = {
```

Comment:
use `some` to instead of `forEach` . if `some` find result will auto interrupt execution , but `forEach` will always execute .

I suggest you use `some` to refactor this code and get better performance.

see mdn : [some](https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Array/some)

Reply:
hmmm, also you can use [find](https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Array/find) enough

Reply:
**** you, Gavin; this part is WIP, review later.

Reply:
hmm, I considered if a `keyword` can match 2 headers. so, I used `foreach`

Reply:
oh thats a bug. if found more than a header matched it only add the last one to `toReturn`.